File: app/src/main/java/social/entourage/android/tour/choice/ChoiceFragment.java
package social.entourage.android.tour.choice;

import android.app.Activity;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.Window;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import social.entourage.android.EntourageApplication;
import social.entourage.android.EntourageComponent;
import social.entourage.android.R;
import social.entourage.android.api.model.map.Tour;

public class ChoiceFragment extends DialogFragment implements ChoiceAdapter.RecyclerViewClickListener {

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @Inject
    ChoicePresenter presenter;

    @BindView(R.id.choice_recycler_view)
    RecyclerView recyclerView;

    private List<Tour> tours;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public static ChoiceFragment newInstance(Tour.Tours tours) {
        ChoiceFragment fragment = new ChoiceFragment();
        Bundle args = new Bundle();
        args.putSerializable(Tour.KEY_TOURS, tours);
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        if (getDialog() != null && getDialog().getWindow() != null) {
            getDialog().getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        }
        View toReturn = inflater.inflate(R.layout.fragment_choice, container, false);
        ButterKnife.bind(this, toReturn);

        Tour.Tours receivedTours = (Tour.Tours) getArguments().getSerializable(Tour.KEY_TOURS);
        if (receivedTours != null) {
            tours = receivedTours.getTours();
        }
        else {
            tours = new ArrayList<>();
        }
        Collections.sort(tours, new Tour.TourComparatorNewToOld());
        initializeView();

        return toReturn;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupComponent(EntourageApplication.get(getActivity()).getEntourageComponent());
    }

    protected void setupComponent(EntourageComponent entourageComponent) {
        DaggerChoiceComponent.builder()
                .entourageComponent(entourageComponent)
                .choiceModule(new ChoiceModule(this))
                .build()
                .inject(this);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getDialog() != null && getDialog().getWindow() != null && getDialog().getWindow().getAttributes() != null) {
            getDialog().getWindow().getAttributes().windowAnimations = R.style.CustomDialogFragmentFade;
        }
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------

    private void initializeView() {
        recyclerView.addItemDecoration(new RecyclerView.ItemDecoration() {
        });
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setAdapter(new ChoiceAdapter(this, tours));
        recyclerView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                recyclerView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                if (recyclerView.getChildCount() > 0) {
                    final View lastCell = recyclerView.getChildAt(recyclerView.getChildCount() - 1);
                    if (lastCell.getBottom() < recyclerView.getHeight()) {
                        final ViewGroup.LayoutParams layoutParams = recyclerView.getLayoutParams();
                        layoutParams.height = lastCell.getBottom();
                        recyclerView.setLayoutParams(layoutParams);
                    }
                }

            }
        });
    }

    private OnChoiceFragmentFinish getOnChoiceFragmentFinish() {
        final Activity activity = getActivity();
        return activity != null ? (OnChoiceFragmentFinish) activity : null;
    }

    @Override
    public void recyclerViewListClicked(Tour tour) {
        if (tour != null && getOnChoiceFragmentFinish() != null) {
            getOnChoiceFragmentFinish().closeChoiceFragment(this, tour);
        }
    }

    // ----------------------------------
    // INNER CLASSE
    // ----------------------------------

    public interface OnChoiceFragmentFinish {
        void closeChoiceFragment(ChoiceFragment fragment, Tour tour);
    }
}

File: app/src/main/java/social/entourage/android/tour/choice/ChoiceFragment.kt
package social.entourage.android.tour.choice

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.view.Window
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import kotlinx.android.synthetic.main.fragment_choice.*
import social.entourage.android.R
import social.entourage.android.api.model.tour.Tour
import social.entourage.android.api.model.tour.Tour.TourComparatorNewToOld
import social.entourage.android.api.model.tour.Tour.Tours
import social.entourage.android.tour.choice.ChoiceAdapter.RecyclerViewClickListener
import java.util.*

class ChoiceFragment : DialogFragment(), RecyclerViewClickListener {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    private val tours: List<Tour> by lazy {
        (requireArguments().getSerializable(Tour.KEY_TOURS) as Tours?)?.tours ?: ArrayList()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialog?.window?.requestFeature(Window.FEATURE_NO_TITLE)
        val toReturn = inflater.inflate(R.layout.fragment_choice, container, false)
        Collections.sort(tours, TourComparatorNewToOld())
        initializeView()
        return toReturn
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        dialog?.window?.attributes?.windowAnimations = R.style.CustomDialogFragmentFade
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------
    private fun initializeView() {
        choice_recycler_view?.let {
            it.addItemDecoration(object : ItemDecoration() {})
            it.layoutManager = LinearLayoutManager(activity)
            it.adapter = ChoiceAdapter(this, tours)
            it.viewTreeObserver.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
                override fun onGlobalLayout() {
                    it.viewTreeObserver.removeOnGlobalLayoutListener(this)
                    if (it.childCount > 0) {
                        val lastCell = it.getChildAt(it.childCount - 1)
                        if (lastCell.bottom < it.height) {
                            val layoutParams = it.layoutParams
                            layoutParams.height = lastCell.bottom
                            it.layoutParams = layoutParams
                        }
                    }
                }
            })
        }
    }

    private val onChoiceFragmentFinish: OnChoiceFragmentFinish?
        get() = activity as? OnChoiceFragmentFinish?

    override fun recyclerViewListClicked(tour: Tour) {
        onChoiceFragmentFinish?.closeChoiceFragment(this, tour)
    }

    // ----------------------------------
    // INNER CLASSE
    // ----------------------------------
    interface OnChoiceFragmentFinish {
        fun closeChoiceFragment(fragment: ChoiceFragment?, tour: Tour?)
    }

    companion object {
        // ----------------------------------
        // LIFECYCLE
        // ----------------------------------
        fun newInstance(tours: Tours?): ChoiceFragment {
            val fragment = ChoiceFragment()
            val args = Bundle()
            args.putSerializable(Tour.KEY_TOURS, tours)
            fragment.arguments = args
            return fragment
        }
    }
}